### PR TITLE
[Balance] Remove candy friendship loss from fainting

### PR DESCRIPTION
**Pull request by Fontbane**
## What are the changes the user will see?

- You no longer lose candy progress (aka candy friendship) when your Pokémon faints

## Why am I making these changes?

- The candy loss from fainting is pretty much anti-QOL, and it's odd to have a way for a player to lose meta progression
- Part of the effort to alleviate the candy grind seen in #4947 
- I previously adjusted this in #2552 

## What are the changes from a developer perspective?

- Removed the block in the `PlayerPokemon.addFriendship` function in `pokemon.ts` that adjusts candy friendship when the `friendship` parameter (the friendship to be added/subtracted) is less than 0
- Because it's no longer used in that else block, the function has been cleaned up a little so that all starter-based code is moved inside the if block (which only runs if the friendship change is positive)

## Screenshots/Videos

A fresh Weedle
Screenshot_977
Weedle after fainting
Screenshot_978
Weedle if it wins instead of faints
Screenshot_979

## Checklist
- [X] **I'm using `beta` as my base branch**
- [X] There is no overlap with another PR?
- [X] The PR is self-contained and cannot be split into smaller PRs?
- [X] Have I provided a clear explanation of the changes?
- [X] Have I tested the changes manually?
- [X] Are all unit tests still passing? (`npm run test`)
  - [ ] Have I created new automated tests (`npm run create-test`) or updated existing tests related to the PR's changes?
- [X] Have I provided screenshots/videos of the changes (if applicable)?
  - [ ] Have I made sure that any UI change works for both UI themes (default and legacy)?

Are there any localization additions or changes? If so:
- [ ] Has a locales PR been created on the [locales](https://github.com/pagefaultgames/pokerogue-locales) repo?
  - [ ] If so, please leave a link to it here: 
- [ ] Has the translation team been contacted for proofreading/translation?

### CODE REVIEW

File: src/field/pokemon.ts [hunk 1]
```diff
@@ -4272,26 +4272,27 @@ export class PlayerPokemon extends Pokemon {
...
+    if (friendship > 0) {
+      const starterSpeciesId = this.species.getRootSpeciesId();
+      const fusionStarterSpeciesId = this.isFusion() && this.fusionSpecies ? this.fusionSpecies.getRootSpeciesId() : 0;
+      const starterData = [
+        this.scene.gameData.starterData[starterSpeciesId],
+        fusionStarterSpeciesId ? this.scene.gameData.starterData[fusionStarterSpeciesId] : null
+      ].filter(d => !!d);
+      const amount = new Utils.IntegerHolder(friendship);
       this.scene.applyModifier(PokemonFriendshipBoosterModifier, true, this, amount);
-      this.scene.applyModifier(PokemonFriendshipBoosterModifier, true, this, starterAmount);
```
Comment (DayKev) [inline src/field/pokemon.ts:4289]: It seems like the `PokemonFriendshipBoosterModifier` is never applied to the candy gain now?

**Comment by DayKev** [comment]
Note: it might be helpful to enable "ignore whitespace" when reviewing the PR.

image

File: src/field/pokemon.ts [hunk 2]
```diff
@@ -4272,26 +4272,27 @@ export class PlayerPokemon extends Pokemon {
...
+      let candyFriendshipMultiplier = CLASSIC_CANDY_FRIENDSHIP_MULTIPLIER;
+      if (this.scene.eventManager.isEventActive()) {
+        candyFriendshipMultiplier *= this.scene.eventManager.getFriendshipMultiplier();
+      }
+      const starterAmount = new Utils.IntegerHolder(Math.floor(friendship * (this.scene.gameMode.isClassic && friendship > 0 ? candyFriendshipMultiplier : 1) / (fusionStarterSpeciesId ? 2 : 1)));
```
Comment (MokaStitcher) [inline src/field/pokemon.ts:4288]:
```suggestion
      const starterAmount = new Utils.NumberHolder(Math.floor(friendship * (this.scene.gameMode.isClassic ? candyFriendshipMultiplier : 1) / (fusionStarterSpeciesId ? 2 : 1)));
```
`friendship > 0` should always be true so it's no longer needed

File: src/field/pokemon.ts [hunk 3]
```diff
@@ -4272,26 +4272,27 @@ export class PlayerPokemon extends Pokemon {
...
 
+      //Add friendship to this PlayerPokemon
       this.friendship = Math.min(this.friendship + amount.value, 255);
       if (this.friendship === 255) {
         this.scene.validateAchv(achvs.MAX_FRIENDSHIP);
       }
+      //Add to candy progress for this mon's starter species and its fused species (if it has one)
```
Comment (MokaStitcher) [inline src/field/pokemon.ts:4295]:
```suggestion
      // Add to candy progress for this mon's starter species and its fused species (if it has one)
```

File: src/field/pokemon.ts [hunk 4]
```diff
@@ -4301,10 +4302,8 @@ export class PlayerPokemon extends Pokemon {
         }
       });
     } else {
-      this.friendship = Math.max(this.friendship + amount.value, 0);
-      for (const sd of starterData) {
-        sd.friendship = Math.max((sd.friendship || 0) + starterAmount.value, 0);
-      }
+      //Lose friendship upon fainting
```
Comment (MokaStitcher) [inline src/field/pokemon.ts:4305]:
```suggestion
      // Lose friendship upon fainting
```

File: src/field/pokemon.ts [hunk 5]
```diff
@@ -4272,26 +4272,27 @@ export class PlayerPokemon extends Pokemon {
   }
 
   addFriendship(friendship: integer): void {
```
Comment (MokaStitcher) [inline src/field/pokemon.ts:4274]:
```suggestion
  addFriendship(friendship: number): void {
```

File: src/field/pokemon.ts [hunk 6]
```diff
@@ -4272,26 +4272,27 @@ export class PlayerPokemon extends Pokemon {
...
-    const starterAmount = new Utils.IntegerHolder(Math.floor(friendship * (this.scene.gameMode.isClassic && friendship > 0 ? candyFriendshipMultiplier : 1) / (fusionStarterSpeciesId ? 2 : 1)));
-    if (amount.value > 0) {
+    if (friendship > 0) {
+      const starterSpeciesId = this.species.getRootSpeciesId();
+      const fusionStarterSpeciesId = this.isFusion() && this.fusionSpecies ? this.fusionSpecies.getRootSpeciesId() : 0;
+      const starterData = [
+        this.scene.gameData.starterData[starterSpeciesId],
+        fusionStarterSpeciesId ? this.scene.gameData.starterData[fusionStarterSpeciesId] : null
+      ].filter(d => !!d);
+      const amount = new Utils.IntegerHolder(friendship);
```
Comment (MokaStitcher) [inline src/field/pokemon.ts:4282]:
```suggestion
      const amount = new Utils.NumberHolder(friendship);
```

File: src/field/pokemon.ts [hunk 7]
```diff
@@ -4272,26 +4272,27 @@ export class PlayerPokemon extends Pokemon {
...
+      }
+      const starterAmount = new Utils.IntegerHolder(Math.floor(friendship * (this.scene.gameMode.isClassic && friendship > 0 ? candyFriendshipMultiplier : 1) / (fusionStarterSpeciesId ? 2 : 1)));
 
+      //Add friendship to this PlayerPokemon
       this.friendship = Math.min(this.friendship + amount.value, 255);
       if (this.friendship === 255) {
         this.scene.validateAchv(achvs.MAX_FRIENDSHIP);
       }
+      //Add to candy progress for this mon's starter species and its fused species (if it has one)
       starterData.forEach((sd: StarterDataEntry, i: integer) => {
```
Comment (MokaStitcher) [inline src/field/pokemon.ts:4296]:
```suggestion
      starterData.forEach((sd: StarterDataEntry, i: number) => {
```

**Comment by star-krieg** [comment]
> just some nits + code cleanup Same issue as DayKev though, is it intended that Soothe Bell does not apply anymore to the candy friendship?

Soothe Bell should still apply to candy friendship (it would defeat the point of putting Soothe Bells to Great)

File: src/field/pokemon.ts (hunk 1 above)
Comment (Fontbane) [reply src/field/pokemon.ts:4289]: My bad, meant to pass amount.value into starterAmount rather than friendship, which would render this line redundant.

**Review by DayKev — APPROVED** [review]

**Review by frutescens — APPROVED** [review]

**Pull request merged**